File: packages/firebase_core/firebase_core/ios/firebase_core/Sources/firebase_core/FLTFirebaseCorePlugin.m
// Copyright 2020 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#if __has_include("include/firebase_core/FLTFirebaseCorePlugin.h")
#import "include/firebase_core/FLTFirebaseCorePlugin.h"
#else
#import "include/FLTFirebaseCorePlugin.h"
#endif

#if __has_include("include/firebase_core/FLTFirebasePluginRegistry.h")
#import "include/firebase_core/FLTFirebasePluginRegistry.h"
#else
#import "include/FLTFirebasePluginRegistry.h"
#endif

#if __has_include("include/firebase_core/messages.g.h")
#import "include/firebase_core/messages.g.h"
#else
#import "include/messages.g.h"
#endif

@implementation FLTFirebaseCorePlugin {
  BOOL _coreInitialized;
}

#pragma mark - FlutterPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
  FLTFirebaseCorePlugin *sharedInstance = [self sharedInstance];
#if TARGET_OS_OSX
#else
  [registrar publish:sharedInstance];
#endif
  FirebaseCoreHostApiSetup(registrar.messenger, sharedInstance);
  FirebaseAppHostApiSetup(registrar.messenger, sharedInstance);
}

// Returns a singleton instance of the Firebase Core plugin.
+ (instancetype)sharedInstance {
  static dispatch_once_t onceToken;
  static FLTFirebaseCorePlugin *instance;

  dispatch_once(&onceToken, ^{
    instance = [[FLTFirebaseCorePlugin alloc] init];
    // Register with the Flutter Firebase plugin registry.
    [[FLTFirebasePluginRegistry sharedInstance] registerFirebasePlugin:instance];

    // Initialize default Firebase app, but only if the plist file options
    // exist.
    //  - If it is missing then there is no default app discovered in Dart and
    //  Dart throws an error.
    //  - Without this the iOS/MacOS app would crash immediately on calling
    //  [FIRApp configure] without
    //    providing helpful context about the crash to the user.
    //
    // Default app exists check is for backwards compatibility of legacy
    // FlutterFire plugins that call [FIRApp configure]; themselves internally.
    FIROptions *options = [FIROptions defaultOptions];
    if (options != nil && [FIRApp allApps][@"__FIRAPP_DEFAULT"] == nil) {
      [FIRApp configureWithOptions:options];
    }
  });

  return instance;
}

static NSMutableDictionary<NSString *, NSString *> *customAuthDomains;

// Initialize static properties

+ (void)initialize {
  if (self == [FLTFirebaseCorePlugin self]) {
    customAuthDomains = [[NSMutableDictionary alloc] init];
  }
}

+ (NSString *)getCustomDomain:(NSString *)appName {
  return customAuthDomains[appName];
}

#pragma mark - Helpers

- (PigeonFirebaseOptions *)optionsFromFIROptions:(FIROptions *)options {
  PigeonFirebaseOptions *pigeonOptions = [PigeonFirebaseOptions alloc];
  pigeonOptions.apiKey = (id)options.APIKey ?: [NSNull null];
  pigeonOptions.appId = (id)options.googleAppID ?: [NSNull null];
  pigeonOptions.messagingSenderId = (id)options.GCMSenderID ?: [NSNull null];
  pigeonOptions.projectId = (id)options.projectID ?: [NSNull null];
  pigeonOptions.databaseURL = (id)options.databaseURL ?: [NSNull null];
  pigeonOptions.storageBucket = (id)options.storageBucket ?: [NSNull null];
  pigeonOptions.deepLinkURLScheme = (id)options.deepLinkURLScheme ?: [NSNull null];
  pigeonOptions.iosBundleId = (id)options.bundleID ?: [NSNull null];
  pigeonOptions.iosClientId = (id)options.clientID ?: [NSNull null];
  pigeonOptions.appGroupId = (id)options.appGroupID ?: [NSNull null];
  return pigeonOptions;
}

- (PigeonInitializeResponse *)initializeResponseFromFIRApp:(FIRApp *)firebaseApp {
  NSString *appNameDart = [FLTFirebasePlugin firebaseAppNameFromIosName:firebaseApp.name];
  PigeonInitializeResponse *response = [PigeonInitializeResponse alloc];
  response.name = appNameDart;
  response.options = [self optionsFromFIROptions:firebaseApp.options];
  response.isAutomaticDataCollectionEnabled = @(firebaseApp.isDataCollectionDefaultEnabled);
  response.pluginConstants =
      [[FLTFirebasePluginRegistry sharedInstance] pluginConstantsForFIRApp:firebaseApp];

  return response;
}

#pragma mark - FLTFirebasePlugin

- (void)didReinitializeFirebaseCore:(void (^)(void))completion {
  completion();
}

- (NSDictionary *_Nonnull)pluginConstantsForFIRApp:(FIRApp *)firebase_app {
  return @{};
}

- (NSString *_Nonnull)firebaseLibraryName {
  return @LIBRARY_NAME;
}

- (NSString *_Nonnull)firebaseLibraryVersion {
  return @LIBRARY_VERSION;
}

- (NSString *_Nonnull)flutterChannelName {
  // The pigeon channel depends on each function
  return @"dev.flutter.pigeon.FirebaseCoreHostApi.initializeApp";
}

#pragma mark - API

- (void)initializeAppAppName:(nonnull NSString *)appName
        initializeAppRequest:(nonnull PigeonFirebaseOptions *)initializeAppRequest
                  completion:(nonnull void (^)(PigeonInitializeResponse *_Nullable,
                                               FlutterError *_Nullable))completion {
  NSString *appNameIos = [FLTFirebasePlugin firebaseAppNameFromDartName:appName];

  if ([FLTFirebasePlugin firebaseAppNamed:appNameIos]) {
    completion([self initializeResponseFromFIRApp:[FLTFirebasePlugin firebaseAppNamed:appNameIos]],
               nil);
    return;
  }

  NSString *appId = initializeAppRequest.appId;
  NSString *messagingSenderId = initializeAppRequest.messagingSenderId;
  FIROptions *options = [[FIROptions alloc] initWithGoogleAppID:appId
                                                    GCMSenderID:messagingSenderId];

  options.APIKey = initializeAppRequest.apiKey;
  options.projectID = initializeAppRequest.projectId;

  // kFirebaseOptionsDatabaseUrl
  if (![initializeAppRequest.databaseURL isEqual:[NSNull null]]) {
    options.databaseURL = initializeAppRequest.databaseURL;
  }

  // kFirebaseOptionsStorageBucket
  if (![options.storageBucket isEqual:[NSNull null]]) {
    options.storageBucket = initializeAppRequest.storageBucket;
  }

  // kFirebaseOptionsDeepLinkURLScheme
  if (![initializeAppRequest.deepLinkURLScheme isEqual:[NSNull null]]) {
    options.deepLinkURLScheme = initializeAppRequest.deepLinkURLScheme;
  }

  // kFirebaseOptionsIosBundleId
  if (![initializeAppRequest.iosBundleId isEqual:[NSNull null]]) {
    options.bundleID = initializeAppRequest.iosBundleId;
  }

  // kFirebaseOptionsIosClientId
  if (![initializeAppRequest.iosClientId isEqual:[NSNull null]]) {
    options.clientID = initializeAppRequest.iosClientId;
  }

  // kFirebaseOptionsAppGroupId
  if (![initializeAppRequest.appGroupId isEqual:[NSNull null]]) {
    options.appGroupID = initializeAppRequest.appGroupId;
  }

  if (initializeAppRequest.authDomain != nil) {
    customAuthDomains[appNameIos] = initializeAppRequest.authDomain;
  }

  [FIRApp configureWithName:appNameIos options:options];

  completion([self initializeResponseFromFIRApp:[FIRApp appNamed:appNameIos]], nil);
}

- (void)initializeCoreWithCompletion:
    (nonnull void (^)(NSArray<PigeonInitializeResponse *> *_Nullable,
                      FlutterError *_Nullable))completion {
  void (^initializeCoreBlock)(void) = ^void() {
    NSDictionary<NSString *, FIRApp *> *firebaseApps = [FIRApp allApps];
    NSMutableArray *firebaseAppsArray = [NSMutableArray arrayWithCapacity:firebaseApps.count];

    for (NSString *appName in firebaseApps) {
      FIRApp *firebaseApp = firebaseApps[appName];
      [firebaseAppsArray addObject:[self initializeResponseFromFIRApp:firebaseApp]];
    }

    completion(firebaseAppsArray, nil);
  };

  if (!_coreInitialized) {
    _coreInitialized = YES;
    initializeCoreBlock();
  } else {
    [[FLTFirebasePluginRegistry sharedInstance] didReinitializeFirebaseCore:initializeCoreBlock];
  }
}

- (void)optionsFromResourceWithCompletion:(nonnull void (^)(PigeonFirebaseOptions *_Nullable,
                                                            FlutterError *_Nullable))completion {
  // Unsupported on iOS/MacOS.
  completion(nil, nil);
}

- (void)deleteAppName:(nonnull NSString *)appName
           completion:(nonnull void (^)(FlutterError *_Nullable))completion {
  FIRApp *firebaseApp = [FLTFirebasePlugin firebaseAppNamed:appName];

  if (firebaseApp) {
    [firebaseApp deleteApp:^(BOOL success) {
      if (success) {
        completion(nil);
      } else {
        completion([FlutterError errorWithCode:@"delete-failed"
                                       message:@"Failed to delete a Firebase app instance."
                                       details:nil]);
      }
    }];
  } else {
    completion(nil);
  }
}

- (void)setAutomaticDataCollectionEnabledAppName:(nonnull NSString *)appName
                                         enabled:(nonnull NSNumber *)enabled
                                      completion:
                                          (nonnull void (^)(FlutterError *_Nullable))completion {
  FIRApp *firebaseApp = [FLTFirebasePlugin firebaseAppNamed:appName];
  if (firebaseApp) {
    [firebaseApp setDataCollectionDefaultEnabled:[enabled boolValue]];
  }

  completion(nil);
}

- (void)setAutomaticResourceManagementEnabledAppName:(nonnull NSString *)appName
                                             enabled:(nonnull NSNumber *)enabled
                                          completion:(nonnull void (^)(FlutterError *_Nullable))
                                                         completion {
  // Unsupported on iOS/MacOS.
  completion(nil);
}

@end

File: packages/firebase_core/firebase_core/ios/firebase_core/Sources/firebase_core/FLTFirebaseCorePlugin.swift
// Copyright 2020 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import FirebaseCore

#if canImport(firebase_core_objc)
  @_exported import firebase_core_objc
#endif

#if canImport(FlutterMacOS)
  import FlutterMacOS
#else
  import Flutter
#endif

extension FlutterError: Error {}

private let kFLTFirebaseCoreChannelName =
  "dev.flutter.pigeon.FirebaseCoreHostApi.initializeApp"

@objc(FLTFirebaseCorePlugin)
public class FLTFirebaseCorePlugin: FLTFirebasePlugin, FlutterPlugin, FLTFirebasePluginProtocol,
  FirebaseCoreHostApi, FirebaseAppHostApi
{
  private var coreInitialized = false

  private static let shared: FLTFirebaseCorePlugin = {
    let instance = FLTFirebaseCorePlugin()
    FLTFirebasePluginRegistry.sharedInstance().register(instance)

    // Initialize default Firebase app, but only if the plist file options exist.
    //  - If it is missing then there is no default app discovered in Dart and
    //    Dart throws an error.
    //  - Without this the iOS/macOS app would crash immediately on calling
    //    FirebaseApp.configure() without providing helpful context about the crash.
    //
    // Default app exists check is for backwards compatibility of legacy
    // FlutterFire plugins that call FirebaseApp.configure() themselves internally.
    if FirebaseOptions.defaultOptions() != nil,
      FirebaseApp.allApps?["__FIRAPP_DEFAULT"] == nil
    {
      FirebaseApp.configure()
    }

    return instance
  }()

  public static func register(with registrar: FlutterPluginRegistrar) {
    let binaryMessenger: FlutterBinaryMessenger

    #if os(macOS)
      binaryMessenger = registrar.messenger
    #elseif os(iOS)
      binaryMessenger = registrar.messenger()
    #endif

    let instance = shared
    #if os(iOS)
      registrar.publish(instance)
    #endif
    FirebaseCoreHostApiSetup.setUp(binaryMessenger: binaryMessenger, api: instance)
    FirebaseAppHostApiSetup.setUp(binaryMessenger: binaryMessenger, api: instance)
  }

  public func didReinitializeFirebaseCore(_ completion: @escaping () -> Void) {
    completion()
  }

  public func pluginConstants(for firebaseApp: FirebaseApp) -> [AnyHashable: Any] {
    [:]
  }

  @objc public func firebaseLibraryName() -> String {
    "flutter-fire-core"
  }

  public func firebaseLibraryVersion() -> String {
    versionNumber
  }

  @objc public func flutterChannelName() -> String {
    kFLTFirebaseCoreChannelName
  }

  private func pigeonOptions(from options: FirebaseOptions) -> CoreFirebaseOptions {
    CoreFirebaseOptions(
      apiKey: options.apiKey ?? "",
      appId: options.googleAppID,
      messagingSenderId: options.gcmSenderID,
      projectId: options.projectID ?? "",
      authDomain: nil,
      databaseURL: options.databaseURL,
      storageBucket: options.storageBucket,
      measurementId: nil,
      trackingId: nil,
      deepLinkURLScheme: nil,
      androidClientId: nil,
      iosClientId: options.clientID,
      iosBundleId: options.bundleID,
      appGroupId: options.appGroupID
    )
  }

  private func pluginConstantsDictionary(from firebaseApp: FirebaseApp) -> [String?: Any?] {
    let raw = FLTFirebasePluginRegistry.sharedInstance().pluginConstants(for: firebaseApp)
    var constants: [String?: Any?] = [:]
    for (key, value) in raw {
      if let stringKey = key as? String {
        constants[stringKey] = value
      }
    }
    return constants
  }

  private func initializeResponse(from firebaseApp: FirebaseApp) -> CoreInitializeResponse {
    return CoreInitializeResponse(
      name: FLTFirebasePlugin.firebaseAppName(fromIosName: firebaseApp.name),
      options: pigeonOptions(from: firebaseApp.options),
      isAutomaticDataCollectionEnabled: firebaseApp.isDataCollectionDefaultEnabled,
      pluginConstants: pluginConstantsDictionary(from: firebaseApp)
    )
  }

  func initializeApp(
    appName: String,
    initializeAppRequest: CoreFirebaseOptions,
    completion: @escaping (Result<CoreInitializeResponse, Error>) -> Void
  ) {
    let appNameIos = FLTFirebasePlugin.firebaseAppName(fromDartName: appName)

    if let existing = FLTFirebasePlugin.firebaseAppNamed(appNameIos) {
      completion(.success(initializeResponse(from: existing)))
      return
    }

    let options = FirebaseOptions(
      googleAppID: initializeAppRequest.appId,
      gcmSenderID: initializeAppRequest.messagingSenderId
    )
    options.apiKey = initializeAppRequest.apiKey
    options.projectID = initializeAppRequest.projectId

    if let databaseURL = initializeAppRequest.databaseURL {
      options.databaseURL = databaseURL
    }
    if let storageBucket = initializeAppRequest.storageBucket {
      options.storageBucket = storageBucket
    }
    if let iosBundleId = initializeAppRequest.iosBundleId {
      options.bundleID = iosBundleId
    }
    if let iosClientId = initializeAppRequest.iosClientId {
      options.clientID = iosClientId
    }
    if let appGroupId = initializeAppRequest.appGroupId {
      options.appGroupID = appGroupId
    }

    if let authDomain = initializeAppRequest.authDomain {
      FLTFirebasePlugin.setCustomAuthDomain(authDomain, forAppName: appNameIos)
    }

    FirebaseApp.configure(name: appNameIos, options: options)

    if let firebaseApp = FirebaseApp.app(name: appNameIos) {
      completion(.success(initializeResponse(from: firebaseApp)))
    } else {
      completion(
        .failure(
          FlutterError(
            code: "initialize-failed",
            message: "Failed to initialize a Firebase app instance.",
            details: nil
          )
        )
      )
    }
  }

  func initializeCore(
    completion: @escaping (Result<[CoreInitializeResponse], Error>) -> Void
  ) {
    let initializeCoreBlock = {
      let firebaseApps = FirebaseApp.allApps ?? [:]
      let responses = firebaseApps.values.map { self.initializeResponse(from: $0) }
      completion(.success(Array(responses)))
    }

    if !coreInitialized {
      coreInitialized = true
      initializeCoreBlock()
    } else {
      FLTFirebasePluginRegistry.sharedInstance().didReinitializeFirebaseCore(initializeCoreBlock)
    }
  }

  func optionsFromResource(
    completion: @escaping (Result<CoreFirebaseOptions, Error>) -> Void
  ) {
    // Unsupported on iOS/macOS. Dart only calls this on Android.
    completion(
      .success(
        CoreFirebaseOptions(
          apiKey: "",
          appId: "",
          messagingSenderId: "",
          projectId: ""
        )
      )
    )
  }

  func delete(
    appName: String,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    guard let firebaseApp = FLTFirebasePlugin.firebaseAppNamed(appName) else {
      completion(.success(()))
      return
    }

    firebaseApp.delete { success in
      if success {
        completion(.success(()))
      } else {
        completion(
          .failure(
            FlutterError(
              code: "delete-failed",
              message: "Failed to delete a Firebase app instance.",
              details: nil
            )
          )
        )
      }
    }
  }

  func setAutomaticDataCollectionEnabled(
    appName: String,
    enabled: Bool,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    if let firebaseApp = FLTFirebasePlugin.firebaseAppNamed(appName) {
      firebaseApp.isDataCollectionDefaultEnabled = enabled
    }
    completion(.success(()))
  }

  func setAutomaticResourceManagementEnabled(
    appName: String,
    enabled: Bool,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    // Unsupported on iOS/macOS.
    completion(.success(()))
  }
}
